index/index.go

```diff
@@ -0,0 +1,393 @@
+// Copyright (C) 2022 CYBERCRYPT
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published
+// by the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+package index
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/gob"
+	"errors"
+	"fmt"
+
+	"github.com/gofrs/uuid"
+	"golang.org/x/crypto/sha3"
+
+	"github.com/cybercryptio/d1-lib/crypto"
+	"github.com/cybercryptio/d1-lib/data"
+	"github.com/cybercryptio/d1-lib/id"
+	"github.com/cybercryptio/d1-lib/io"
+	"github.com/cybercryptio/d1-lib/key"
+)
+
+// The length of the master key.
+const MasterKeyLength = 32
+
+// Error returned if a master key with invalid length is used.
+var ErrInvalidMasterKeyLength = fmt.Errorf("invalid key length, accepted key length is %d bytes", MasterKeyLength)
+
+// Error returned if the caller cannot be authenticated by the Identity Provider.
+var ErrNotAuthenticated = errors.New("user not authenticated")
+
+// Error returned if a user tries to access data they are not authorized for.
+var ErrNotAuthorized = errors.New("user not authorized")
+
+type SecureIndex struct {
+	keyProvider key.Provider
```

Comment:
We don't actually use the key provider after construction, so we can omit this field.